src/ivector/agglomerative-clustering.cc

```diff
@@ -0,0 +1,132 @@
+        (active_clusters_.find(j) != active_clusters_.end()))
+      MergeClusters(i, j);
+  }
+
+  std::vector<int32> new_assignments(num_points_);
+  int32 label_id = 0;
+  std::set<int32>::iterator it;
+  // Iterate through the clusters and assign all utterances within the cluster
+  // an ID label unique to the cluster. This is the final output and frees up
+  // the cluster memory accordingly.
+  for (it = active_clusters_.begin(); it != active_clusters_.end(); ++it) {
+    ++label_id;
+    AhcCluster *cluster = clusters_map_[*it];
+    std::vector<int32>::iterator utt_it;
+    for (utt_it = cluster->utt_ids.begin();
+         utt_it != cluster->utt_ids.end(); ++utt_it)
+      new_assignments[*utt_it] = label_id;
+    delete cluster;
+  }
+  assignments_->swap(new_assignments);
+}
+
+BaseFloat AgglomerativeClusterer::ScoreLookup(int32 i, int32 j) {
+  if (i < j)
+    return cluster_score_map_[std::make_pair(i, j)];
+  else
+    return cluster_score_map_[std::make_pair(j, i)];
+}
+
+void AgglomerativeClusterer::Initialize() {
+  KALDI_ASSERT(num_clusters_ != 0);
+  for (int32 i = 0; i < num_points_; i++) {
+    // create an initial cluster of size 1 for each point
+    std::vector<int32> ids;
+    ids.push_back(i);
+    AhcCluster *c = new AhcCluster(++count_, -1, -1, ids);
+    clusters_map_[count_] = c;
+    active_clusters_.insert(count_);
+
+    // propagate the queue with all pairs from the score matrix
+    for (int32 j = i+1; j < num_clusters_; j++) {
+      BaseFloat score = scores_(i,j);
+      cluster_score_map_[std::make_pair(i+1, j+1)] = score;
+      if (score <= thresh_)
+        queue_.push(std::make_pair(score,
+            std::make_pair(static_cast<uint16>(i+1),
+                           static_cast<uint16>(j+1))));
+    }
+  }
+}
+
+void AgglomerativeClusterer::MergeClusters(int32 i, int32 j) {
+  AhcCluster *clust1 = clusters_map_[i];
+  AhcCluster *clust2 = clusters_map_[j];
+  // For memory efficiency, the first cluster is updated to contain the new
+  // merged cluster information, and the second cluster is later deleted.
+  clust1->id = ++count_;
+  clust1->parent1 = i;
+  clust1->parent2 = j;
+  clust1->size += clust2->size;
+  clust1->utt_ids.insert(clust1->utt_ids.end(), clust2->utt_ids.begin(),
+                         clust2->utt_ids.end());
+  // Remove the merged clusters from the list of active clusters.
+  active_clusters_.erase(i);
+  active_clusters_.erase(j);
+  // Update the queue with all the new scores involving the new cluster
+  std::set<int32>::iterator it;
+  for (it = active_clusters_.begin(); it != active_clusters_.end(); ++it) {
+    // The new score is the sum of the score of the new cluster's parents
+    BaseFloat new_score = ScoreLookup(*it, i) + ScoreLookup(*it, j);
+    cluster_score_map_[std::make_pair(*it, count_)] = new_score;
+    BaseFloat norm = clust1->size * (clusters_map_[*it])->size;
+    if (new_score / norm <= thresh_)
```

Comment:
It seems these "scores" are really "distances". So change the name accordingly.

Reply:
Since @mmaciej2 will be editing this anyway, it'd be nice to change ScoreLookup to GetDistance... Most Kaldi code uses "Get" for accessors.